### PR TITLE
Adding cache and specialization to func decorator

### LINKED ISSUE #867: Can't reuse `dpex.func` device functions  with different signatures
This used to work with `numba_dpex<=0.18.1` but fails with `>=0.19`:

```python
import numba_dpex as dpex
import dpctl.tensor as dpt
import numpy as np


@dpex.func
def g(array_in, idx, const):
    array_in[idx] = const

@dpex.kernel
def kernel_a(array_in):
    idx = dpex.get_global_id(0)
    g(array_in, idx, np.int64(0))

@dpex.kernel
def kernel_b(array_in):
    idx = dpex.get_global_id(0)
    g(array_in, idx, np.int32(0))  # NB: call with inputs of different types than in kernel_a

   
dtype = np.float32
size = 16
array_in = dpt.zeros(sh=(size,), dtype=dtype)

kernel_a[size, size](array_in)
kernel_b[size, size](array_in)
```

with `numba_dpex>=0.19` this snippet gives the following exception:

```
<...traceback elided...>
LoweringError: Failed in dpex_nopython mode pipeline (step: Custom Lowerer with auto-offload support)
No definition for lowering <function get_global_id at 0x7f47c2edf700>(uint32,) -> int64

File "<ipython-input-14-e32b2bd5f6a0>", line 17:
def kernel_b(array_in):
    idx = dpex.get_global_id(0)
    ^

During: lowering "idx = call $4load_method.1($const6.2, func=$4load_method.1, args=[Var($const6.2, <ipython-input-14-e32b2bd5f6a0>:17)], kws=(), vararg=None, varkwarg=None, target=None)" at <ipython-input-14-e32b2bd5f6a0> (17)
```

There's a workaround that consists in ensuring that in both kernels the device function is at different memory locations with a different name, e.g.:

```python
import numba_dpex as dpex
import dpctl.tensor as dpt
import numpy as np


def make_g():
    @dpex.func
    def g(array_in, idx, const):
        array_in[idx] = const
    return g

g = make_g()
@dpex.kernel
def kernel_a(array_in):
    idx = dpex.get_global_id(0)
    g(array_in, idx, np.int64(0))

g_ = make_g()
@dpex.kernel
def kernel_b(array_in):
    idx = dpex.get_global_id(0)
    g_(array_in, idx, np.int32(0))

   
dtype = np.float32
size = 16
array_in = dpt.zeros(sh=(size,), dtype=dtype)

kernel_a[size, size](array_in)
kernel_b[size, size](array_in)
```

### PR DESCRIPTION
**Pull request by chudur-budur**
Fixes #867 and #861. 
- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [x] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: numba_dpex/core/kernel_interface/func.py [hunk 1]
```diff
@@ -69,37 +73,85 @@ def compile(self, args):
...
-                debug=self.debug,
+                debug=self._debug,
             )
             func = cres.library.get_function(cres.fndesc.llvm_func_name)
             cres.target_context.mark_ocl_device(func)
-            first_definition = not self._compileinfos
-            self._compileinfos[args] = cres
             libs = [cres.library]
 
-            if first_definition:
```
Comment (chudur-budur) [inline numba_dpex/core/kernel_interface/func.py:89]: This might have solved the double `__call__` problem? @diptorupd

File: notes.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
+1. Specialized function gets compiled when the file is loaded
```
Comment (diptorupd) [inline notes.md:1]: Please remove temp file.

File: numba_dpex/core/kernel_interface/func.py [hunk 3]
```diff
@@ -6,100 +6,217 @@
...
-        typing_context=dpex_target.typing_context,
-        args=args,
-        is_kernel=False,
-        debug=debug,
-    )
-    func = cres.library.get_function(cres.fndesc.llvm_func_name)
-    cres.target_context.mark_ocl_device(func)
-    devfn = DpexFunction(cres)
+class DpexFunction(object):
+    """Class to materialize dpex function"""
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/func.py:20]: The description can be improved. How about:

"Helper class to eager compile a specialized `numba_dpex.func` decorated Python function into a LLVM function with `spir_func` calling convention."

File: numba_dpex/core/kernel_interface/func.py [hunk 4]
```diff
@@ -6,100 +6,217 @@
...
-                raise AssertionError("No keyword arguments allowed.")
-            return dft.compile(args)
-
-    dpex_target.typing_context.insert_user_function(dft, _function_template)
-    return dft
+        return cres
 
 
 class DpexFunctionTemplate(object):
     """Unmaterialized dpex function"""
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/func.py:62]: On similar lines as `DpexFunction`, please improve the docstring.

File: numba_dpex/core/kernel_interface/func.py [hunk 5]
```diff
@@ -6,100 +6,217 @@
...
+
+    @property
+    def cache_hits(self):
+        """Cache hit count accessor"""
+        return self._cache_hits
 
     def compile(self, args):
-        """Compile a dpex.func decorated Python function with the given
-        argument types.
+        """Compile a dpex.func decorated Python function
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/func.py:101]: s/"dpex.func"/"numba_dpex.func"/g

File: numba_dpex/core/kernel_interface/func.py [hunk 6]
```diff
@@ -6,100 +6,217 @@
...
 
-            if first_definition:
-                # First definition
-                cres.target_context.insert_user_function(
-                    self, cres.fndesc, libs
-                )
-            else:
-                cres.target_context.add_user_function(self, cres.fndesc, libs)
+            cres.target_context.insert_user_function(self, cres.fndesc, libs)
+            # cres.target_context.add_user_function(self, cres.fndesc, libs)
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/func.py:142]: Please remove commented out line.

File: numba_dpex/core/kernel_interface/func.py [hunk 7]
```diff
@@ -6,100 +6,217 @@
...
+            self._cache.put(key, cres)
+        return cres.signature
 
-        else:
-            cres = self._compileinfos[args]
 
-        return cres.signature
+def compile_func(pyfunc, signature, debug=None):
+    """AI is creating summary for compile_func
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/func.py:148]: What does the docstring mean?

File: numba_dpex/core/kernel_interface/func.py [hunk 8]
```diff
@@ -6,100 +6,217 @@
...
+        c = devfn.compile(arg_types, return_types)
+        cres.append(c)
+
+    class _function_template(ConcreteTemplate):
+        unsafe_casting = False
+        exact_match_required = True
+        key = devfn
+        cases = [c.signature for c in cres]
+
+    # TODO: If context is only one copy, just make it global
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/func.py:181]: Is the TODO valid?

File: numba_dpex/decorators.py [hunk 9]
```diff
@@ -94,39 +94,46 @@ def _specialized_kernel_dispatcher(pyfunc):
...
-    )
-
-    def _wrapped(pyfunc):
-        return compile_func(pyfunc, restype, argtypes, debug=debug)
-
-    return _wrapped
-
+def func(func_or_sig=None, debug=None, enable_cache=True):
+    """
+    TODO: make similar to kernel version and when the signature is a list
```
Comment (diptorupd) [inline numba_dpex/decorators.py:99]: Can be removed?

Comment (diptorupd) [reply numba_dpex/decorators.py:99]: And we need a proper docstring.

File: numba_dpex/decorators.py [hunk 10]
```diff
@@ -94,39 +94,46 @@ def _specialized_kernel_dispatcher(pyfunc):
...
+    if func_or_sig is None:
+        return _func_autojit_wrapper(debug=debug)
+    elif isinstance(func_or_sig, str):
+        raise NotImplementedError(
+            "Specifying signatures as string is not yet supported by numba-dpex"
+        )
+    elif isinstance(func_or_sig, list) or sigutils.is_signature(func_or_sig):
+        # String signatures are not supported as passing usm_ndarray type as
+        # a string is not possible. Numba's sigutils relies on the type being
+        # available in Numba's types.__dpct__ and dpex types are not registered
```
Comment (diptorupd) [inline numba_dpex/decorators.py:118]: Typo `types.__dpct__` should be `types.__dict__`

File: numba_dpex/examples/kernel/device_func.py [hunk 11]
```diff
@@ -30,24 +45,90 @@ def a_kernel_function(a, b):
...
+
+    print("Using device ...")
+    print(a.device)
+
+    print("A=", a)
+    a_kernel_function_int32_float32[N](a, b)
+    print("B=", b)
+
+    # this will fail, since int64 is not in
+    # the signature list: [int32(int32), float32(float32)]
```
Comment (diptorupd) [inline numba_dpex/examples/kernel/device_func.py:116]: try except is required or else CI will complain.

File: test-func-specialize.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
+# SPDX-FileCopyrightText: 2020 - 2022 Intel Corporation
```
Comment (diptorupd) [inline test-func-specialize.py:1]: remove the file.

File: numba_dpex/decorators.py [hunk 13]
```diff
@@ -69,6 +68,7 @@ def _kernel_dispatcher(pyfunc, sigs=None):
         if not isinstance(func_or_sig, list):
             func_or_sig = [func_or_sig]
 
+        # TODO: This function is 99% same as _kernel_dispatcher()
```
Comment (diptorupd) [inline numba_dpex/decorators.py:71]: TODO should be removed.

File: numba_dpex/core/kernel_interface/func.py [hunk 14]
```diff
@@ -6,100 +6,224 @@
...
+            cres.target_context.insert_user_function(self, cres.fndesc, libs)
+            self._cache.put(key, cres)
+        return cres.signature
 
-        else:
-            cres = self._compileinfos[args]
 
-        return cres.signature
+def compile_func(pyfunc, signature, debug=None):
+    """Compile a DpexFunction
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/func.py:158]: Docstring needs fixing, How about:

"Compiles a specialized `numba_dpex.func` decorated function to native binary library function and returns the library wrapped inside a :class:`numba_dpex.core.kernel_interface.func.DpexFunction` object."

File: numba_dpex/core/kernel_interface/func.py [hunk 15]
```diff
@@ -6,100 +6,224 @@
...
+            debug=self._debug,
+        )
+        func = cres.library.get_function(cres.fndesc.llvm_func_name)
+        cres.target_context.mark_ocl_device(func)
+
+        return cres
 
 
 class DpexFunctionTemplate(object):
     """Unmaterialized dpex function"""
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/func.py:73]: Docstring should be updated to:

"Helper class to JIT compile an unspecialized numba_dpex.func
 decorated Python function into an LLVM function with spir_func calling convention."

**Review by diptorupd — APPROVED** [review]

**Pull request merged**